File: Core/Features/Tags/CreateTag.cs
namespace Core.Features.Tags;

public static class CreateTag
{
    public record Command(
        int TenantCode,
        string Code,
        string TagCategoryCode) : IRequest;

    public class Handler(
        ITenantRepository tenantRepository,
        ITagCategoryRepository tagCategoryRepository,
        ITagRepository tagRepository) : IRequestHandler<Command>
    {
        public async Task Handle(Command request, CancellationToken cancellationToken)
        {
            if (!await tenantRepository.ExistsByCodeAsync(request.TenantCode, cancellationToken))
            {
                throw new TenantNotFoundException(request.TenantCode);
            }

            if (!await tagCategoryRepository.ExistsByCodeAsync(request.TagCategoryCode, cancellationToken))
            {
                throw new TagCategoryNotFoundException(request.TagCategoryCode);
            }

            var tag = new Tag(
                request.TenantCode,
                request.Code,
                request.TagCategoryCode);

            await tagRepository.CreateAsync(tag, cancellationToken);
        }
    }
}

File: backoffice/src/server/queries/tags/mutations/create-tag.ts
import { fetchGraphQL } from "@/utils";

export async function createTagMutation(
  tenantCode: number,
  tagCategoryCode: string,
  code: string
): Promise<string> {
  const query = `
    mutation CreateTag($tenantCode: Int!, $tagCategoryCode: String!, $code: String!) {
        createTag(command: {
            tenantCode: $tenantCode,
            tagCategoryCode: $tagCategoryCode,
            code: $code
        }) {
            code
        }
    }
`;

  const tag = await fetchGraphQL<{ createTag: { code: string } }>(query, {
    tenantCode,
    tagCategoryCode,
    code,
  });

  return tag.createTag.code;
}
